File: 8_String_to_Integer_atoi/solution.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.lstrip()
        if not s:
            return 0
            
        num, sign, i = 0, 1, 0
        if s[0] in "+-":
            sign = 1 if s[0] == "+" else -1
            i = 1

        while i < len(s) and s[i].isdigit():
            num = num * 10 + (ord(s[i]) - ord("0"))
            i += 1

        INT_MAX, INT_MIN = 2**31 - 1, -(2**31)
        val = sign * num

        if val > INT_MAX:
            return INT_MAX
        if val < INT_MIN:
            return INT_MIN
        return val

File: 8_String_to_Integer_atoi/solution.c
int myAtoi(char* s) {
    long long result = 0;
    int sign = 1;
    while (*s == ' ') { s++; }

    if (*s == '+' || *s == '-') { if (*s == '-') sign = -1; s++; }

    while (isdigit((unsigned char)*s)) {
        int digit = *s - '0';
        if (result > INT_MAX / 10 || (result == INT_MAX / 10 && digit > (sign == 1 ? 7 : 8))) { return (sign == 1) ? INT_MAX : INT_MIN; }
        result = result * 10 + digit;
        s++;
    }
    return (int)(sign * result);
}
